compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantTypeChecker.java

```diff
@@ -0,0 +1,2847 @@
+            varRefExpr.symbol = new BVarSymbol(0, true, varName,
+                    names.originalNameFromIdNode(varRefExpr.variableName),
+                    data.env.enclPkg.symbol.pkgID, varRefExpr.getBType(), data.env.scope.owner,
+                    varRefExpr.pos, VIRTUAL);
+
+            data.resultType = varRefExpr.getBType();
+            return;
+        }
+
+        Name compUnitName = typeChecker.getCurrentCompUnit(varRefExpr);
+        varRefExpr.pkgSymbol =
+                symResolver.resolvePrefixSymbol(data.env, names.fromIdNode(varRefExpr.pkgAlias), compUnitName);
+        if (varRefExpr.pkgSymbol == symTable.notFoundSymbol) {
+            varRefExpr.symbol = symTable.notFoundSymbol;
+            dlog.error(varRefExpr.pos, DiagnosticErrorCode.UNDEFINED_MODULE, varRefExpr.pkgAlias);
+        }
+
+        if (varRefExpr.pkgSymbol != symTable.notFoundSymbol) {
+            BSymbol symbol =
+                    getSymbolOfVarRef(varRefExpr.pos, data.env, names.fromIdNode(varRefExpr.pkgAlias), varName, data);
+
+            if (symbol == symTable.notFoundSymbol) {
+                data.resultType = symTable.semanticError;
+                return;
+            }
+
+            if ((symbol.tag & SymTag.CONSTANT) == SymTag.CONSTANT) {
+                // Check whether the referenced expr is a constant.
+                BConstantSymbol constSymbol = (BConstantSymbol) symbol;
+                varRefExpr.symbol = constSymbol;
+                actualType = symbol.type;
+            } else {
+                varRefExpr.symbol = symbol;
+                dlog.error(varRefExpr.pos, DiagnosticErrorCode.EXPRESSION_IS_NOT_A_CONSTANT_EXPRESSION);
+            }
+        }
+
+        data.resultType = types.checkType(varRefExpr, actualType, data.expType);
+    }
+
+    public void visit(BLangRecordLiteral recordLiteral, AnalyzerData data) {
+        BType expType = data.expType;
+        int expTypeTag = Types.getReferredType(expType).tag;
+
+        if (expTypeTag == TypeTags.NONE || expTypeTag == TypeTags.READONLY) {
+            data.resultType = validateMapTypeAndInferredType(recordLiteral, expType, expType, data);
+            return;
+        }
+        data.resultType = checkMappingConstructorCompatibility(data.expType, recordLiteral, data);
+    }
+
+    @Override
+    public void visit(BLangBinaryExpr binaryExpr, AnalyzerData data) {
+        BType expType = data.expType;
+
+        data.compoundExprCount++;
+        BType lhsType = checkConstExpr(binaryExpr.lhsExpr, expType, data);
+        data.compoundExprCount--;
+
+        data.compoundExprCount++;
+        BType rhsType = checkConstExpr(binaryExpr.rhsExpr, expType, data);
+        data.compoundExprCount--;
+
+        Location pos = binaryExpr.pos;
+
+        if (lhsType == symTable.semanticError || rhsType == symTable.semanticError) {
+            data.resultType = symTable.semanticError;
+            return;
+        } else {
```

Comment:
`else` is not needed because we return in `if`

Reply:
Updated 70dc2c1266fafd2aedf3400af1f3bc6d778de232